### ISSUE TITLE
openssl 1.0.21 - RSA key exponent not odd

### ISSUE THREAD

**Issue by flynn1973** [openssl/openssl#3949]
this problem originates from within cfengine linked against the mentioned openssl version. (thread on the cfengine mailing list: https://groups.google.com/forum/#!topic/help-cfengine/-hmnIz0fA4Q).

**Comment by richsalz** [comment]
I read the thread and looked at the openssl code.  I don't see how it can be a bug in openssl, as opposed to strangeness in your local configuration.  But let's keep this issue open in case more info comes in.

**Comment by flynn1973** [comment]
here es the spec file used to build the rpm package. is there someting in the configure options that may cause the problems?

`
%ifos aix6.1 || %ifos aix7.1
%global p7build 1
%else
%global p7build 0
%endif

%define openssldir /var/ssl-rpm

%define majorver 1.0.2

Summary: Secure Sockets Layer and cryptography libraries and tools
Name: openssl
Version: %{majorver}l
%if %{p7build}
Release: 1.p7
%else
Release: 1
%endif
Source0: ftp://ftp.openssl.org/source/%{name}-%{version}.tar.gz
Source1: ftp://ftp.openssl.org/source/%{name}-%{version}.tar.gz.asc
Source2: ftp://ftp.openssl.org/source/%{name}-%{version}.tar.gz.sha1
Source3: ftp://ftp.openssl.org/source/%{name}-%{version}.tar.gz.sha256
Source4: libcrypto.so.0.9.7-aix32
Source5: libcrypto.so.0.9.7-aix64
Source6: libssl.so.0.9.7-aix32
Source7: libssl.so.0.9.7-aix64
Source8: libcrypto.so.0.9.8-aix32
Source9: libcrypto.so.0.9.8-aix64
Source10: libssl.so.0.9.8-aix32
Source11: libssl.so.0.9.8-aix64
Source12: libcrypto.so.1.0.0-aix32
Source13: libcrypto.so.1.0.0-aix64
Source14: libssl.so.1.0.0-aix32
Source15: libssl.so.1.0.0-aix64
Source16: libcrypto.so.1.0.1-aix32
Source17: libcrypto.so.1.0.1-aix64
Source18: libssl.so.1.0.1-aix32
Source19: libssl.so.1.0.1-aix64
Patch0: %{name}-%{version}-Configure.patch
License: OpenSSL License
Group: System Environment/Libraries
Provides: SSL
URL: http://www.openssl.org/
BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
Obsoletes: openssl64
%define _prefix64 %{_prefix}/64

%define _libdir64 %{_prefix}/lib64

BuildRequires: make

%if %{p7build}
%ifos aix6.1
Requires: AIX-rpm >= 6.1.8.0
%endif
%ifos aix7.1
Requires: AIX-rpm >= 7.1.3.0
%endif
%endif

%description
The OpenSSL Project is a collaborative effort to develop a robust,
commercial-grade, fully featured, and Open Source toolkit implementing the
Secure Sockets Layer (SSL v2/v3) and Transport Layer Security (TLS v1)
protocols as well as a full-strength general purpose cryptography library.
The project is managed by a worldwide community of volunteers that use the
Internet to communicate, plan, and develop the OpenSSL tookit and its related
documentation.

OpenSSL is based on the excellent SSLeay library developed from Eric A. Young
and Tim J. Hudson.  OpenSSL is licensed under the OpenSSL License, included in
this package.

This package contains the base OpenSSL cryptography and SSL/TLS libraries and
tools.

You should also install a pseudo-random number generator such as EGD or prngd
if you do not have a /dev/urandom and /dev/random.

The library is available as 32-bit and 64-bit.


%package devel
Summary: Secure Sockets Layer and cryptography static libraries and headers
Group: Development/Libraries
Requires: %{name} = %{version}
Obsoletes: openssl64-devel

%description devel
The OpenSSL Project is a collaborative effort to develop a robust,
commercial-grade, fully featured, and Open Source toolkit implementing the
Secure Sockets Layer (SSL v2/v3) and Transport Layer Security (TLS v1)
protocols as well as a full-strength general purpose cryptography library.
The project is managed by a worldwide community of volunteers that use the
Internet to communicate, plan, and develop the OpenSSL tookit and its related
documentation.

OpenSSL is based on the excellent SSLeay library developed from Eric A. Young
and Tim J. Hudson.  OpenSSL is licensed under the OpenSSL License, included in
this package.

This package contains the the OpenSSL cryptography and SSL/TLS static libraries
and header files required when developing applications.

If you are compiling a 32-bit program, no special compiler options are
needed.

If you are compiling a 64-bit program, you have to compile and link your
application with "xlc_r -q64" or "gcc -maix64".


%package doc
Summary: OpenSSL miscellaneous files
Group: Documentation
Requires: %{name} = %{version}
Obsoletes: openssl64-doc

%description doc
The OpenSSL Project is a collaborative effort to develop a robust,
commercial-grade, fully featured, and Open Source toolkit implementing the
Secure Sockets Layer (SSL v2/v3) and Transport Layer Security (TLS v1)
protocols as well as a full-strength general purpose cryptography library.
The project is managed by a worldwide community of volunteers that use the
Internet to communicate, plan, and develop the OpenSSL tookit and its related
documentation.

OpenSSL is based on the excellent SSLeay library developed from Eric A. Young
and Tim J. Hudson.  OpenSSL is licensed under the OpenSSL License, included in
this package.

This package contains the the OpenSSL cryptography and SSL/TLS extra
documentation and POD files from which the man pages were produced.


%prep
%setup -q
%patch0


%build
/usr/bin/perl util/perlpath.pl /usr/bin/perl

export CC="xlc_r"

export RM="/usr/bin/rm -f"

# first build the 64-bit version
export OBJECT_MODE=64
./Configure \
    no-idea no-mdc2 no-rc5 \
    -DSSL_ALLOW_ADH \
    threads \
    no-shared \
    zlib-dynamic \
    --prefix=%{_prefix} \
    --openssldir=%{openssldir} \
%if %{p7build}
    aix64p7-xlc_r
%else
    no-asm \
    aix64-xlc_r
%endif

gmake depend
gmake build_libs

CreateExportList -X64 libcrypto.exp libcrypto.a
${CC} -qmkshrobj libcrypto.a -o libcrypto.so.%{majorver} -bE:libcrypto.exp
rm -f libcrypto.a libcrypto.exp
/usr/bin/ar -rv libcrypto.a libcrypto.so.%{majorver}

CreateExportList -X64 libssl.exp libssl.a
${CC} -qmkshrobj libssl.a -o libssl.so.%{majorver} -bE:libssl.exp -L. -lcrypto -blibpath:/opt/freeware/lib:/usr/lib:/lib
rm -f libssl.a libssl.exp
/usr/bin/ar -rv libssl.a libssl.so.%{majorver}

gmake

# apps/openssl still has the wrong shared library paths, we have to relink it
# this is an ugly hack:
# In order to get the correct shared libraries linked they must reside before
# linking in /opt/freeware/lib, e.g. we have to copy them manually there!
rm -f apps/openssl
gmake LIBSSL="-L/opt/freeware/lib -lssl" LIBCRYPTO="-L/opt/freeware/lib -lcrypto"

mkdir -p 64
mv apps/openssl libcrypto.so.%{majorver} libssl.so.%{majorver} 64/
gmake clean
slibclean

####---------------------

# now build the 32-bit version
export OBJECT_MODE=32
./Configure \
    no-idea no-mdc2 no-rc5 \
    -DSSL_ALLOW_ADH \
    threads \
    no-shared \
    zlib-dynamic \
    --prefix=%{_prefix} \
    --openssldir=%{openssldir} \
%if %{p7build}
    aixp7-xlc_r
%else
    no-asm \
    aix-xlc_r
%endif

gmake depend
gmake build_libs

CreateExportList -X32 libcrypto.exp libcrypto.a
${CC} -qmkshrobj libcrypto.a -o libcrypto.so.%{majorver} -bE:libcrypto.exp
rm -f libcrypto.a libcrypto.exp
/usr/bin/ar -rv libcrypto.a libcrypto.so.%{majorver}

CreateExportList -X32 libssl.exp libssl.a
${CC} -qmkshrobj libssl.a -o libssl.so.%{majorver} -bE:libssl.exp -L. -lcrypto -blibpath:/opt/freeware/lib:/usr/lib:/lib
rm -f libssl.a libssl.exp
/usr/bin/ar -rv libssl.a libssl.so.%{majorver}

gmake
slibclean

# add the 64-bit shared objects to the shared library containing already the
# 32-bit shared objects
/usr/bin/ar -X64 -q libcrypto.a 64/libcrypto.so.%{majorver}
/usr/bin/ar -X64 -q libssl.a 64/libssl.so.%{majorver}

# Add the older 0.9.7l, 0.9.8zh, 1.0.0t and 1.0.1t shared members for compatibility with older apps
# (make sure they're set for LOADONLY with 'strip -e')
cp %{SOURCE4} libcrypto.so.0.9.7
/usr/bin/strip -X32 -e libcrypto.so.0.9.7
/usr/bin/ar -X32 -q libcrypto.a libcrypto.so.0.9.7
cp %{SOURCE8} libcrypto.so.0.9.8
/usr/bin/strip -X32 -e libcrypto.so.0.9.8
/usr/bin/ar -X32 -q libcrypto.a libcrypto.so.0.9.8
cp %{SOURCE12} libcrypto.so.1.0.0
/usr/bin/strip -X32 -e libcrypto.so.1.0.0
/usr/bin/ar -X32 -q libcrypto.a libcrypto.so.1.0.0
cp %{SOURCE16} libcrypto.so.1.0.1
/usr/bin/strip -X32 -e libcrypto.so.1.0.1
/usr/bin/ar -X32 -q libcrypto.a libcrypto.so.1.0.1

cp %{SOURCE5} libcrypto.so.0.9.7
/usr/bin/strip -X64 -e libcrypto.so.0.9.7
/usr/bin/ar -X64 -q libcrypto.a libcrypto.so.0.9.7
cp %{SOURCE9} libcrypto.so.0.9.8
/usr/bin/strip -X64 -e libcrypto.so.0.9.8
/usr/bin/ar -X64 -q libcrypto.a libcrypto.so.0.9.8
cp %{SOURCE13} libcrypto.so.1.0.0
/usr/bin/strip -X64 -e libcrypto.so.1.0.0
/usr/bin/ar -X64 -q libcrypto.a libcrypto.so.1.0.0
cp %{SOURCE17} libcrypto.so.1.0.1
/usr/bin/strip -X64 -e libcrypto.so.1.0.1
/usr/bin/ar -X64 -q libcrypto.a libcrypto.so.1.0.1

cp %{SOURCE6} libssl.so.0.9.7
/usr/bin/strip -X32 -e libssl.so.0.9.7
/usr/bin/ar -X32 -q libssl.a libssl.so.0.9.7
cp %{SOURCE10} libssl.so.0.9.8
/usr/bin/strip -X32 -e libssl.so.0.9.8
/usr/bin/ar -X32 -q libssl.a libssl.so.0.9.8
cp %{SOURCE14} libssl.so.1.0.0
/usr/bin/strip -X32 -e libssl.so.1.0.0
/usr/bin/ar -X32 -q libssl.a libssl.so.1.0.0
cp %{SOURCE18} libssl.so.1.0.1
/usr/bin/strip -X32 -e libssl.so.1.0.1
/usr/bin/ar -X32 -q libssl.a libssl.so.1.0.1

cp %{SOURCE7} libssl.so.0.9.7
/usr/bin/strip -X64 -e libssl.so.0.9.7
/usr/bin/ar -X64 -q libssl.a libssl.so.0.9.7
cp %{SOURCE11} libssl.so.0.9.8
/usr/bin/strip -X64 -e libssl.so.0.9.8
/usr/bin/ar -X64 -q libssl.a libssl.so.0.9.8
cp %{SOURCE15} libssl.so.1.0.0
/usr/bin/strip -X64 -e libssl.so.1.0.0
/usr/bin/ar -X64 -q libssl.a libssl.so.1.0.0
cp %{SOURCE19} libssl.so.1.0.1
/usr/bin/strip -X64 -e libssl.so.1.0.1
/usr/bin/ar -X64 -q libssl.a libssl.so.1.0.1

# apps/openssl still has the wrong shared library paths, we have to relink it
# this is an ugly hack:
# In order to get the correct shared libraries linked they must reside before
# linking in /opt/freeware/lib, e.g. we have to copy them manually there!
rm -f apps/openssl
gmake LIBSSL="-L/opt/freeware/lib -lssl" LIBCRYPTO="-L/opt/freeware/lib -lcrypto"


%install
[ "${RPM_BUILD_ROOT}" != "/" ] && rm -rf ${RPM_BUILD_ROOT}
gmake INSTALL_PREFIX=${RPM_BUILD_ROOT} MANDIR=%{_mandir} install

cp 64/openssl ${RPM_BUILD_ROOT}%{_bindir}/openssl64
chmod 755 ${RPM_BUILD_ROOT}%{_bindir}/openssl64

strip -X32_64 ${RPM_BUILD_ROOT}%{_bindir}/* || :

# Make backwards-compatibility symlink to ssleay
ln -s %{_bindir}/openssl ${RPM_BUILD_ROOT}%{_bindir}/ssleay
ln -s %{_bindir}/openssl64 ${RPM_BUILD_ROOT}%{_bindir}/ssleay64

# add 32-bit run-time-linker libraries
cd ${RPM_BUILD_ROOT}%{_libdir}
/usr/bin/ar -X32 -x libcrypto.a
/usr/bin/ar -X32 -x libssl.a
ln -s libcrypto.so.%{majorver} libcrypto.so
ln -s libssl.so.%{majorver} libssl.so

# add 64-bit run-time-linker libraries
mkdir -p ${RPM_BUILD_ROOT}%{_libdir64}/
cd ${RPM_BUILD_ROOT}%{_libdir64}
/usr/bin/ar -X64 -x ../lib/libcrypto.a
/usr/bin/ar -X64 -x ../lib/libssl.a
ln -s libcrypto.so.%{majorver} libcrypto.so
ln -s libssl.so.%{majorver} libssl.so

(
  cd ${RPM_BUILD_ROOT}
  for dir in bin lib lib64 include
  do
    mkdir -p usr/linux/${dir}
    cd usr/linux/${dir}
    ln -sf ../../..%{_prefix}/${dir}/* .
    cd -
  done

# change these to be full-path links
  cd usr/linux/lib
  rm ./libcrypto.* ./libssl.*
  ln -sf %{_libdir}/libcrypto.a .
  ln -sf %{_libdir}/libcrypto.so .
  ln -sf %{_libdir}/libcrypto.so.%{majorver} .
  ln -sf %{_libdir}/libssl.a .
  ln -sf %{_libdir}/libssl.so .
  ln -sf %{_libdir}/libssl.so.%{majorver} .

  cd ../lib64
  rm ./libcrypto.* ./libssl.*
  ln -sf %{_libdir64}/libcrypto.so .
  ln -sf %{_libdir64}/libcrypto.so.%{majorver} .
  ln -sf %{_libdir64}/libssl.so .
  ln -sf %{_libdir64}/libssl.so.%{majorver} .

  cd -
)

# add links for 64-bit library members
(
  mkdir -p ${RPM_BUILD_ROOT}%{_prefix64}/lib
  cd ${RPM_BUILD_ROOT}%{_prefix64}/lib
  ln -s ../../lib/*.a .
)

# rename ${RPM_BUILD_ROOT}%{_mandir}/man3/threads.3 as it might conflict with
# man3/threads.3 from perl
cd ${RPM_BUILD_ROOT}%{_mandir}/man3
mv -f threads.3 threads.3ssl


%clean
[ "${RPM_BUILD_ROOT}" != "/" ] && rm -rf ${RPM_BUILD_ROOT}


%files
%defattr(-,root,system,-)
%doc CHANGES CHANGES.SSLeay LICENSE NEWS README
%dir %{openssldir}
%config %attr(0644,root,system) %{openssldir}/openssl.cnf
%dir %attr(0755,root,system) %{openssldir}/certs
%dir %attr(0755,root,system) %{openssldir}/misc
%dir %attr(0750,root,system) %{openssldir}/private
%{openssldir}/misc/*
%{_bindir}/*
%{_libdir}/*.a
%{_libdir}/*.so*
%{_libdir64}/*.so*
%dir %attr(755,bin,bin) %{_prefix64}
%dir %attr(755,bin,bin) %{_prefix64}/lib
%{_prefix64}/lib/*.a
%{_mandir}/man1/*
%{_mandir}/man5/*
%{_mandir}/man7/*
/usr/linux/bin/*
/usr/linux/lib/*.a
/usr/linux/lib/*.so*
/usr/linux/lib64/*.so*


%files devel
%defattr(-,root,system)
%doc CHANGES CHANGES.SSLeay LICENSE NEWS README
%{_includedir}/*
%{_libdir}/pkgconfig/*
/usr/linux/include/*


%files doc
%defattr(-,root,system)
%doc CHANGES CHANGES.SSLeay LICENSE NEWS README
%doc doc
%{_mandir}/man3/*

`

patch used to add back some missing aix targets

`
*** ./Configure.ORIG    Tue Feb  5 21:47:24 2013
--- ./Configure Tue Feb  5 21:49:34 2013
***************
*** 465,472 ****
  "aix64-gcc","gcc:-maix64 -O -DB_ENDIAN::-pthread:AIX::SIXTY_FOUR_BIT_LONG RC4_CHAR:${ppc64_asm}:aix64:dlfcn:aix-shared::-maix64 -shared -Wl,-G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X64",
  # Below targets assume AIX 5. Idea is to effectively disregard $OBJECT_MODE
  # at build time. $OBJECT_MODE is respected at ./config stage!
! "aix-cc",   "cc:-q32 -O -DB_ENDIAN -qmaxmem=16384 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::BN_LLONG RC4_CHAR:${ppc32_asm}:aix32:dlfcn:aix-shared::-q32 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 32",
! "aix64-cc", "cc:-q64 -O -DB_ENDIAN -qmaxmem=16384 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::SIXTY_FOUR_BIT_LONG RC4_CHAR:${ppc64_asm}:aix64:dlfcn:aix-shared::-q64 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 64",

  #
  # Cray T90 and similar (SDSC)
--- 465,474 ----
  "aix64-gcc","gcc:-maix64 -O -DB_ENDIAN::-pthread:AIX::SIXTY_FOUR_BIT_LONG RC4_CHAR:${ppc64_asm}:aix64:dlfcn:aix-shared::-maix64 -shared -Wl,-G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X64",
  # Below targets assume AIX 5. Idea is to effectively disregard $OBJECT_MODE
  # at build time. $OBJECT_MODE is respected at ./config stage!
! "aix-xlc_r",     "xlc_r:-q32 -O3 -qstrict -DB_ENDIAN -qmaxmem=65536 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::BN_LLONG RC4_CHAR:${ppc32_asm}:aix32:dlfcn:aix-shared::-q32 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 32",
! "aix64-xlc_r",   "xlc_r:-q64 -O3 -qstrict -DB_ENDIAN -qmaxmem=65536 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::SIXTY_FOUR_BIT_LONG RC4_CHAR:${ppc64_asm}:aix64:dlfcn:aix-shared::-q64 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 64",
! "aixp7-xlc_r",   "xlc_r:-q32 -O3 -qstrict -qarch=pwr7 -qtune=pwr7 -DB_ENDIAN -qmaxmem=-1 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::BN_LLONG RC4_CHAR:${ppc32_asm}:aix32:dlfcn:aix-shared::-q32 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 32",
! "aix64p7-xlc_r", "xlc_r:-q64 -O3 -qstrict -qarch=pwr7 -qtune=pwr7 -DB_ENDIAN -qmaxmem=-1 -qro -qroconst::-qthreaded -D_THREAD_SAFE:AIX::SIXTY_FOUR_BIT_LONG RC4_CHAR:${ppc64_asm}:aix64:dlfcn:aix-shared::-q64 -G:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)::-X 64",

  #
  # Cray T90 and similar (SDSC)

`

**Comment by richsalz** [comment]
No, the code should NEVER do an even modulus.  Something is strange.

**Comment by flynn1973** [comment]
fiddled around with it a bit more.

configured as follows ( ./Configure aix64-xlc_r threads shared no-asm zlib-dynamic no-idea no-mdc2 no-rc5 -DSSL_ALLOW_ADH), gmake test fails with this error -> 

```
1 handshakes of 256 bytes done
test sslv2/sslv3 via BIO pair
Available compression methods:
  1: zlib compression
TLSv1.2, cipher TLSv1/SSLv3 ECDHE-RSA-AES256-GCM-SHA384, 2048 bit RSA
1 handshakes of 256 bytes done
test dtlsv1
Available compression methods:
  1: zlib compression
ERROR IN STARTUP
DTLSv1, cipher TLSv1/SSLv3 ECDHE-RSA-AES256-SHA, 2048 bit RSA
1 handshakes of 256 bytes done
Makefile:304: recipe for target 'test_ssl' failed
gmake[1]: *** [test_ssl] Error 1
gmake[1]: Leaving directory '/opt/freeware/src/packages/BUILD/openssl-1.0.2l/test'
Makefile:472: recipe for target 'tests' failed
gmake: *** [tests] Error 2

```

anyway i'm out of options here, any hints where i should look/test are highly appreciated.

**Comment by bernd-edlinger** [comment]
Looks like a compiler bug to me.
I would try lowering the compiler optimization level.

However you should definitely remove the zlib-dynamic because having it opens
the way for a major a security threat: see CRIME attack.

Also please remove -DSSL_ALLOW_ADH is no longer needed since OpenSSL 0.9.5 at least

**Comment by flynn1973** [comment]
with "./Configure aix64-cc --prefix=/usr/local/itsvbuild/64 --openssldir=/usr/local/itsvbuild/64/ssl shared" and a rebuild of cfengine against this version the error is gone. i have not the slightest idea why but it is. the only difference is it was built without rpmbuild which is ODD (pun intended).
however, as there will be no more groundbreaking insights, this thread can be closed.
